modules/scodec/shared/src/main/scala/io/circe/scodec/package.scala

```diff
@@ -2,22 +2,74 @@ package io.circe
 
 import _root_.scodec.bits.{BitVector, ByteVector}
 
-package object scodec {
+private[circe] trait BitVectorCodec {
+  import scala.collection.immutable.StringOps
+  import scala.language.implicitConversions
+
   implicit final val decodeBitVector: Decoder[BitVector] =
     Decoder.instance { c =>
       Decoder.decodeString(c) match {
         case Right(str) =>
-          BitVector.fromBase64Descriptive(str) match {
-            case r @ Right(_) => r.asInstanceOf[Decoder.Result[BitVector]]
-            case Left(err) => Left(DecodingFailure(err, c.history))
+          numberOfCharsToIgnore(str) match {
+            case Some(toIgnore) =>
+              BitVector.fromBase64Descriptive(str.drop(1)) match {
+                case Left(err) => Left(DecodingFailure(err, c.history))
+                case Right(r) =>
+                  val res = r.take(r.size - toIgnore)
+                  Right(res)
+              }
+            case None =>
+              Left(DecodingFailure("Incorrect format of BitVector field", c.history))
           }
         case l @ Left(_) => l.asInstanceOf[Decoder.Result[BitVector]]
       }
     }
 
+  /**
+    * For serialization of BitVector we use base64. scodec's implementation of `toBase64` adds padding to 8 bits.
+    * That's not desired in our case and to preserve original BitVector length we prefix base64 representation with
+    * simple one byte long "header" which stands for number of bits of data in the last byte of base64 representation.
+    *
+    * Examples:
+    * encodeBitVector(bin"101")
+    * res: io.circe.Json = "3oA==" // starts with "3' because BitVector is of size 3
+    *
+    * encodeBitVector(bin"")
+    * res: io.circe.Json = "0"
+    *
+    * encodeBitVector(bin"11001100")
+    * res: io.circe.Json = "8zA=="
+    *
+    */
   implicit final val encodeBitVector: Encoder[BitVector] =
-    Encoder.encodeString.contramap(_.toBase64)
+    Encoder.encodeString.contramap { bv =>
+      significantBitsInLastByte(bv).toString + bv.toBase64
+    }
+
+  @inline private implicit def augmentString(x: String): StringOps = new StringOps(x)
+
+  private def significantBitsInLastByte(bitVector: BitVector): Byte = {
+    if (bitVector.isEmpty) {
+      0
+    } else {
+      val mod = (bitVector.size % 8).toByte
+      if (mod == 0) 8 else mod
+    }
+  }
+
+  private def numberOfCharsToIgnore(input: String): Option[Int] = {
```

Comment:
This function may return `None` in two little bit different cases. Since they're both signalled with the same value it will result in the same error message. I thought it's ok for sake of simplicity but I am open for suggestions.